core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -583,31 +560,67 @@ private[master] class Master(
         for (pos <- 0 until numUsable) {
-          if (assigned(pos) > 0) {
-            val exec = app.addExecutor(usableWorkers(pos), assigned(pos))
-            launchExecutor(usableWorkers(pos), exec)
-            app.state = ApplicationState.RUNNING
-          }
+          allocateWorkerResourceToExecutors(app, assigned(pos), usableWorkers(pos))
         }
       }
     } else {
-      // Pack each app into as few nodes as possible until we've assigned all its cores
+      // Pack each app into as few workers as possible until we've assigned all its cores
       for (worker <- workers if worker.coresFree > 0 && worker.state == WorkerState.ALIVE) {
         for (app <- waitingApps if app.coresLeft > 0) {
-          if (canUse(app, worker)) {
-            val coresToUse = math.min(worker.coresFree, app.coresLeft)
-            if (coresToUse > 0) {
-              val exec = app.addExecutor(worker, coresToUse)
-              launchExecutor(worker, exec)
-              app.state = ApplicationState.RUNNING
-            }
-          }
+          allocateWorkerResourceToExecutors(app, app.coresLeft, worker)
         }
       }
     }
   }
 
-  private def launchExecutor(worker: WorkerInfo, exec: ExecutorDesc) {
+  /**
+   * allocate resources in a certain worker to one or more executors
+   * @param app the info of the application which the executors belong to
+   * @param coresToAllocate cores on this worker to be allocated to this application
+   * @param worker the worker info
+   */
+  private def allocateWorkerResourceToExecutors(
+      app: ApplicationInfo,
+      coresToAllocate: Int,
+      worker: WorkerInfo): Unit = {
+    if (canUse(app, worker)) {
+      val memoryPerExecutor = app.desc.memoryPerExecutorMB
+      val coresPerExecutor = app.desc.coresPerExecutor.getOrElse(coresToAllocate)
+      var coresLeft = coresToAllocate
+      while (coresLeft >= coresPerExecutor && worker.memoryFree >= memoryPerExecutor) {
+        val exec = app.addExecutor(worker, coresPerExecutor)
+        coresLeft -= coresPerExecutor
+        launchExecutor(worker, exec)
+        app.state = ApplicationState.RUNNING
+      }
+    }
+  }
+
+  private def startDriversOnWorkers(): Unit = {
+    val shuffledWorkers = Random.shuffle(workers) // Randomization helps balance drivers
+    for (worker <- shuffledWorkers if worker.state == WorkerState.ALIVE) {
+      for (driver <- waitingDrivers) {
+        if (worker.memoryFree >= driver.desc.mem && worker.coresFree >= driver.desc.cores) {
+          launchDriver(worker, driver)
+          waitingDrivers -= driver
+        }
+      }
+    }
+  }
+
+  /**
+   * Schedule the currently available resources among waiting apps. This method will be called
+   * every time a new app joins or resource availability changes.
+   */
+  private def schedule(): Unit = {
+    if (state != RecoveryState.ALIVE) { return }
+    // start in-cluster drivers, they take strict precedence over applications
```

Comment:
I would just replace this as `// Drivers take strict precedence over executors` and remove the comment in L619, since it doesn't convey any information and doesn't help readability.